modules/core/src/test/java/org/apache/ignite/internal/processors/security/cluster/ClusterStatePermissionTest.java

```diff
@@ -0,0 +1,344 @@
+import org.apache.ignite.internal.processors.security.impl.TestSecurityData;
+import org.apache.ignite.internal.processors.security.impl.TestSecurityPluginProvider;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.plugin.security.SecurityCredentials;
+import org.apache.ignite.plugin.security.SecurityCredentialsBasicProvider;
+import org.apache.ignite.plugin.security.SecurityException;
+import org.apache.ignite.plugin.security.SecurityPermission;
+import org.apache.ignite.plugin.security.SecurityPermissionSetBuilder;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.apache.ignite.cluster.ClusterState.ACTIVE;
+import static org.apache.ignite.cluster.ClusterState.ACTIVE_READ_ONLY;
+import static org.apache.ignite.cluster.ClusterState.INACTIVE;
+import static org.apache.ignite.internal.util.lang.GridFunc.asList;
+import static org.apache.ignite.plugin.security.SecurityPermission.ADMIN_CLUSTER_STATE;
+import static org.apache.ignite.plugin.security.SecurityPermission.JOIN_AS_SERVER;
+import static org.apache.ignite.plugin.security.SecurityPermissionSetBuilder.create;
+import static org.apache.ignite.testframework.GridTestUtils.assertThrowsAnyCause;
+import static org.apache.ignite.testframework.GridTestUtils.cartesianProduct;
+
+/**
+ * Tests permissions of cluster state change.
+ */
+@RunWith(Parameterized.class)
+public class ClusterStatePermissionTest extends AbstractSecurityTest {
+    /** */
+    private static final SecurityPermission[] EMPTY_PERMS = new SecurityPermission[0];
+
+    /** */
+    private SecurityPermission[] permissions = EMPTY_PERMS;
+
+    /** From-client flag parameter. */
+    @Parameterized.Parameter
+    public Initiator initiator;
+
+    /** Persistence flag parameter. */
+    @Parameterized.Parameter(1)
+    public boolean persistence;
+
+    /** @return Test parameters. */
+    @Parameterized.Parameters(name = "nodeType={0}, persistence={1}")
+    public static Collection<?> parameters() {
+        return cartesianProduct(asList(Initiator.values()), asList(false, true));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String instanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(instanceName);
+
+        if (!cfg.isClientMode()) {
+            cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(persistence)));
+        }
+
+        SecurityPermission[] srvPerms = EMPTY_PERMS;
+        SecurityPermission[] clientPerms = EMPTY_PERMS;
+
+        if (initiator == Initiator.SERVER)
```

Comment:
It choses where to apply the required permissions. As an example `permissions = F.asArray(ADMIN_CLUSTER_STATE);` It might be set to client/server/ Or a thin client.

Reply:
Yeah, but now it seems that srvPerms = clientPerms = permissions since ` srvPerms = F.concat(srvPerms, permissions);` appends empty permissions to `srvPerms` now.

Reply:
Aha, got it. I concat empty permissions. That is senseless, yeah. Fixed.